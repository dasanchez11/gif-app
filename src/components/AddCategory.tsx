import { ChangeEvent, FormEvent, useState } from "react";

export interface AddCategoryProps {
  onNewCategory: Function;
}

export const AddCategory = ({ onNewCategory }: AddCategoryProps) => {
  const [inputValue, setInputValue] = useState<string>("");

  const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
    setInputValue(e.target.value);
  };

  const handleSubmit = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const value = inputValue.trim();
    if (value.length <= 1) return;
    onNewCategory(value);
    setInputValue("");
  };

  return (
    <form action="" onSubmit={handleSubmit} aria-label="form">
      <input
        value={inputValue}
        onChange={handleChange}
        type="text"
        placeholder="Search gifs and tap enter"
      />
    </form>
  );
};
