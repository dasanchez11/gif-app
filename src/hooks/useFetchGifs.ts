import { useEffect, useState } from "react";
import { GifImage, getGifs } from "../helpers/getGifs";

export const useFetchGifs = (category: string) => {
  const [images, setImages] = useState<GifImage[]>([]);
  const [loading, setLoading] = useState<boolean>(true);

  const getImages = async () => {
    const newImages = await getGifs(category);
    setImages(newImages);
    setLoading(false);
  };

  useEffect(() => {
    getImages();
  }, []);

  return { images, loading };
};
