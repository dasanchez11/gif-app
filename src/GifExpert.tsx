import { useState } from "react";
import { AddCategory, GifGrid } from "./components";

export const GifExpert = () => {
  const [categories, setCategories] = useState<string[]>([]);

  const onAddCategory = (name: string) => {
    if (categories.includes(name)) return;
    setCategories((state) => [name, ...state]);
  };

  return (
    <>
      <AddCategory onNewCategory={onAddCategory} />
      <h1>GifExpert</h1>
      {categories.map((category) => (
        <GifGrid key={category} category={category} />
      ))}
    </>
  );
};
