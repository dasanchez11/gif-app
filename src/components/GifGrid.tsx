import { GifGridItem } from "./GifGridItem";
import { useFetchGifs } from "../hooks/useFetchGifs";

export const GifGrid = ({ category }: { category: string }) => {
  const { images, loading } = useFetchGifs(category);

  return (
    <>
      <h3>{category}</h3>
      <div className="card-grid">
        {loading && <h2>Loading...</h2>}
        {!loading &&
          images.map((image) => <GifGridItem key={image.id} {...image} />)}
      </div>
    </>
  );
};
