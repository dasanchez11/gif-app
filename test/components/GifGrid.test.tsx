import { render } from "@testing-library/react";
import { GifGrid } from "../../src/components";
import * as fetch from "../../src/hooks/useFetchGifs";

describe("Test <Gifgrid]/>", () => {
  const cateory = "category";
  test("should should display loading initially", () => {
    const returnValue = {
      loading: true,
      images: [],
    };

    jest.spyOn(fetch, "useFetchGifs").mockReturnValueOnce(returnValue);
    const loadingMessage = "Loading...";
    const { container, getByRole } = render(<GifGrid category={cateory} />);
    const loadingElement = getByRole("heading", { level: 2 });
    const categoryElement = getByRole("heading", { level: 3 });
    expect(loadingElement.textContent).toEqual(loadingMessage);
    expect(categoryElement.textContent).toEqual(cateory);
    expect(container).toMatchSnapshot();
  });

  test("should display images", () => {
    const mockImages = [
      {
        id: "abce",
        title: "title",
        url: "urlpage",
      },
      {
        id: "def",
        title: "title2",
        url: "urlpage2",
      },
    ];
    const returnValue = {
      loading: false,
      images: mockImages,
    };
    jest.spyOn(fetch, "useFetchGifs").mockReturnValueOnce(returnValue);
    const { container, getAllByRole } = render(<GifGrid category={cateory} />);
    const cards = getAllByRole("img");
    expect(cards.length).toEqual(2);
  });
});
