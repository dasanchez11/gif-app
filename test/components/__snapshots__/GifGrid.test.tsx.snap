// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <Gifgrid]/> should should display loading initially 1`] = `
<div>
  <h3>
    category
  </h3>
  <div
    class="card-grid"
  >
    <h2>
      Loading...
    </h2>
  </div>
</div>
`;
